### ISSUE TITLE
[Bug]: Milvus crashes if search with expr in a large id list

### ISSUE THREAD

**Issue by ufuk23** [milvus-io/milvus#15536]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.0.0
- Deployment mode(standalone or cluster): standalone
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus
- OS(Ubuntu or CentOS): ubuntu
- CPU/Memory: 4 / 16GB
- GPU: 
- Others:
```

### Current Behavior

I used Milvus 1.0.0 before and want to use 2.0.0. Honestly, while considering resource usage, I am disappointed with 2.0.0. I have a 160.000 vector that is 2048 float values. It must be 160.000 * 2048 * 8 bytes = 2.6 GB, but when I check the docker volume size, it seems 17 GB disk size. There is a big difference. RAM usage is also huge, it's about 4-5 GB for 160K vectors.

I use Milvus 2.0.0 with default configurations, but Milvus 1.0.0 is better than 2.0.0 while considering resource usage. Do I miss something? Can anyone explain what the reason for the difference is?

### Expected Behavior

Normal disk and ram usage

**Comment by xiaofan-luan** [comment]
> ### Is there an existing issue for this?
> * [x]  I have searched the existing issues
> 
> ### Environment
> ```gfm
> - Milvus version: 2.0.0
> - Deployment mode(standalone or cluster): standalone
> - SDK version(e.g. pymilvus v2.0.0rc2): pymilvus
> - OS(Ubuntu or CentOS): ubuntu
> - CPU/Memory: 4 / 16GB
> - GPU: 
> - Others:
> ```
> 
> ### Current Behavior
> I used Milvus 1.0.0 before and want to use 2.0.0. Honestly, while considering resource usage, I am disappointed with 2.0.0. I have a 160.000 vector that is 2048 float values. It must be 160.000 * 2048 * 8 bytes = 2.6 GB, but when I check the docker volume size, it seems 17 GB disk size. There is a big difference. RAM usage is also huge, it's about 4-5 GB for 160K vectors.
> 
> I use Milvus 2.0.0 with default configurations, but Milvus 1.0.0 is better than 2.0.0 while considering resource usage. Do I miss something? Can anyone explain what the reason for the difference is?
> 
> ### Expected Behavior
> Normal disk and ram usage
> 
> ### Steps To Reproduce
> _No response_
> 
> ### Anything else?
> _No response_

Hi @ufuk23, yes milvus 1.0.0 is better on resource usage, especially on disk size.
the reason is:
1) milvus insertion message is stored in rocksmq, and will be purged after a few days.
2) milvus store data on minio, which has 3 replicas 
3) the storage format of milvus 2.0  is more complicated and possibility to do further extend.
4) we stored index as well as original, which should incur 2x of the storage size.
The storage should be much cheaper than memory. so we thought it won't be a huge problem, there are a few things you may want to do in case you want to optimize storage:

1) check data expired time, change it to a smaller value such as 1 day.
2) implement compression on the binlogs, which hopefully would decrease your your datasize.

for memory consumption, we are doing optimization on it but it is a lot bit difficult, for the following reasons:
1) Golang is a gc language, and it maintain some memory as a buffer
2) We have some memory copy issues waiting to be solved.
3) Based on what index you are using, some of the index such as HNSW take much more memory than the original data. If you are using FAISS together with PQ and PCA

**Comment by xiaofan-luan** [comment]
We are working on memory usage and performance optimization, if anyone has any ideas or clues please let me know

**Comment by yanliang567** [comment]
@ufuk23 please check the answers above, feel free to let us know if any more questions

**Comment by ufuk23** [comment]
@yanliang567, @xiaofan-luan, thanks for the answers. I wonder about how to decrease disk usage. I deleted some insert_log files thinking they were unnecessary logs, but the data was affected and could never be searched again. Then I learned that I shouldn't do it again :) I think I can manipulate the milvus.yaml config file to remove log files for a short period. Which parameter is used to do it?

The memory usage is crucial for our case, we have vectors each of them is 2048 float, and if 160K vectors mean 4GB, 1M vectors mean 24 GB RAM, 5M mean 120 GB RAM, and so on. So I am not sure to use Milvus 2.0.0 and don't know if Milvus 3.0 has some architectural changes to force me to change codes again. I am not sure whether I have to reindex the whole data due to the next versions of Milvus. Perhaps, I better use Milvus 1.0.0?

**Comment by yanliang567** [comment]
@ufuk23 thank you for your feedback. We really understand that the memory usage is crucial for everyone. Did you try any indexes for your vectors? For examples, if using ivf_sq8, the memory usage shall theoretically decrease to ~30%.

**Comment by xiaofan-luan** [comment]
> @yanliang567, @xiaofan-luan, thanks for the answers. I wonder about how to decrease disk usage. I deleted some insert_log files thinking they were unnecessary logs, but the data was affected and could never be searched again. Then I learned that I shouldn't do it again :) I think I can manipulate the milvus.yaml config file to remove log files for a short period. Which parameter is used to do it?
> 
> The memory usage is crucial for our case, we have vectors each of them is 2048 float, and if 160K vectors mean 4GB, 1M vectors mean 24 GB RAM, 5M mean 120 GB RAM, and so on. So I am not sure to use Milvus 2.0.0 and don't know if Milvus 3.0 has some architectural changes to force me to change codes again. I am not sure whether I have to reindex the whole data due to the next versions of Milvus. Perhaps, I better use Milvus 1.0.0?

Hi uf, one thing you could do is to list all the directory and check which on consumes most of the data, then I will give more advice about how to decrease datasize.

For the memory consumption, use SQ8 or PQ will significantly decrease the memory usage, please give a try.

Finally, we will try our best to keep Milvus 2.0 compatible and sorry for the trouble between 1.0 and 2.0. We changed the coding language from Cpp to Go for faster data iteration

**Comment by LoveEachDay** [comment]
> @yanliang567, @xiaofan-luan, thanks for the answers. I wonder about how to decrease disk usage. I deleted some insert_log files thinking they were unnecessary logs, but the data was affected and could never be searched again. Then I learned that I shouldn't do it again :) I think I can manipulate the milvus.yaml config file to remove log files for a short period. Which parameter is used to do it?

These are settings for message queue of milvus standalone :
```
rocksmq:
  path: /var/lib/milvus/rdb_data # The path where the message is stored in rocksmq
  rocksmqPageSize: 2147483648 # 2 GB, 2 * 1024 * 1024 * 1024 bytes, The size of each page of messages in rocksmq
  retentionTimeInMinutes: 10080 # 7 days, 7 * 24 * 60 minutes, The retention time of the message in rocksmq.
  retentionSizeInMB: 8192 # 8 GB, 8 * 1024 MB, The retention size of the message in rocksmq.
```
You can modify those `retentionTimeInMinutes` and `retentionSizeInMB` two options to decrease disk usage.

**Comment by ufuk23** [comment]
@LoveEachDay Thanks. I don't know why the retentionTime is set for 7 days as default, isn't it just an insert_log message? I changed the defaults for log configs in my 'milvus.yaml' file which is pointed to as a docker volume file.

@yanliang567, @xiaofan-luan I used the IVF_SQ8 and it seems better than IVF_FLAT. The disk usage is reduced by half. Any disadvantage with ivf_sq8? We assume that the entity count can be 5M in our case. Is there any performance issue about speed using IVF_SQ8?

**Comment by ufuk23** [comment]
I tried 110.000 entities that contain 2048 float vectors with IVF_SQ8. When I use the advanced search with a vector and pk list (containing 100.000 ids) Milvus standalone crashes with the following log.

![milvus2](https://user-images.githubusercontent.com/8513827/154030308-925775c1-a0c4-46d5-828f-390bf76ccbf6.png)


github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80a10)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80a10)
        /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55
goroutine 95778 [select]:
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80a80)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80a80)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55
goroutine 95779 [semacquire]:
sync.runtime_Semacquire(0xc00230e0a8)
	/usr/local/go/src/runtime/sema.go:56 +0x45
sync.(*WaitGroup).Wait(0xc00230e0a0)
	/usr/local/go/src/sync/waitgroup.go:130 +0x65
github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).bufMsgPackToChannel(0xc001978fc0)
	/go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:803 +0x20b
created by github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).Start
	/go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:751 +0x6e
goroutine 95780 [select]:
github.com/milvus-io/milvus/internal/msgstream.(*mqMsgStream).Consume(0xc001978fc0, 0x0)
	/go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:480 +0xdf
github.com/milvus-io/milvus/internal/util/flowgraph.(*InputNode).Operate(0xc001498c30, 0x0, 0x0, 0x0, 0x355f3c0, 0xc002834840, 0x0)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/input_node.go:65 +0x4c
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80930)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:100 +0x214
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55
goroutine 95781 [semacquire]:
sync.runtime_Semacquire(0xc00230e168)
	/usr/local/go/src/runtime/sema.go:56 +0x45
sync.(*WaitGroup).Wait(0xc00230e160)
	/usr/local/go/src/sync/waitgroup.go:130 +0x65
github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).bufMsgPackToChannel(0xc001979440)	/go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:803 +0x20b
created by github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).Start
	/go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:751 +0x6e
goroutine 95782 [select]:
github.com/milvus-io/milvus/internal/msgstream.(*mqMsgStream).Consume(0xc001979440, 0x0)
	/go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:480 +0xdf
github.com/milvus-io/milvus/internal/util/flowgraph.(*InputNode).Operate(0xc001499020, 0x0, 0x0, 0x0, 0x355f3c0, 0xc00204ecc0, 0x0)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/input_node.go:65 +0x4c
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80bd0)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:100 +0x214
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55
goroutine 95783 [select]:
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80c40)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80c40)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55
goroutine 95784 [select]:
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80cb0)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80cb0)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55
goroutine 95785 [select]:
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80d20)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b
github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80d20)
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556
created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start
	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55
rax    0x0
rbx    0x7fc6f37fbf50
rcx    0x7fc7d9b66fb7
rdx    0x0
rdi    0x2
rsi    0x7fc6f37fbce0
rbp    0x7fc6f37fc050
rsp    0x7fc6f37fbce0
r8     0x0
r9     0x7fc6f37fbce0
r10    0x8
r11    0x246
r12    0x7fc6f37fbf50
r13    0x1000
r14    0x0
r15    0x30
rip    0x7fc7d9b66fb7
rflags 0x246
cs     0x33
fs     0x0
gs     0x0

**Comment by xiaofan-luan** [comment]
> I tried 110.000 entities that contain 2048 float vectors with IVF_SQ8. When I use the advanced search with a vector and pk list (containing 100.000 ids) Milvus standalone crashes with the following log.
> 
> ![milvus2](https://user-images.githubusercontent.com/8513827/154030308-925775c1-a0c4-46d5-828f-390bf76ccbf6.png)
> 
> github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80a10) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80a10) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55 goroutine 95778 [select]: github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80a80) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80a80) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55 goroutine 95779 [semacquire]: sync.runtime_Semacquire(0xc00230e0a8) /usr/local/go/src/runtime/sema.go:56 +0x45 sync.(*WaitGroup).Wait(0xc00230e0a0) /usr/local/go/src/sync/waitgroup.go:130 +0x65 github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).bufMsgPackToChannel(0xc001978fc0) /go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:803 +0x20b created by github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).Start /go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:751 +0x6e goroutine 95780 [select]: github.com/milvus-io/milvus/internal/msgstream.(*mqMsgStream).Consume(0xc001978fc0, 0x0) /go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:480 +0xdf github.com/milvus-io/milvus/internal/util/flowgraph.(*InputNode).Operate(0xc001498c30, 0x0, 0x0, 0x0, 0x355f3c0, 0xc002834840, 0x0) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/input_node.go:65 +0x4c github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80930) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:100 +0x214 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55 goroutine 95781 [semacquire]: sync.runtime_Semacquire(0xc00230e168) /usr/local/go/src/runtime/sema.go:56 +0x45 sync.(*WaitGroup).Wait(0xc00230e160) /usr/local/go/src/sync/waitgroup.go:130 +0x65 github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).bufMsgPackToChannel(0xc001979440) /go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:803 +0x20b created by github.com/milvus-io/milvus/internal/msgstream.(*MqTtMsgStream).Start /go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:751 +0x6e goroutine 95782 [select]: github.com/milvus-io/milvus/internal/msgstream.(*mqMsgStream).Consume(0xc001979440, 0x0) /go/src/github.com/milvus-io/milvus/internal/msgstream/mq_msgstream.go:480 +0xdf github.com/milvus-io/milvus/internal/util/flowgraph.(*InputNode).Operate(0xc001499020, 0x0, 0x0, 0x0, 0x355f3c0, 0xc00204ecc0, 0x0) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/input_node.go:65 +0x4c github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80bd0) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:100 +0x214 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55 goroutine 95783 [select]: github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80c40) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80c40) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55 goroutine 95784 [select]: github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80cb0) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80cb0) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55 goroutine 95785 [select]: github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).collectInputMessages(0xc001c80d20) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:157 +0x14b github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001c80d20) /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:96 +0x556 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start /go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:68 +0x55 rax 0x0 rbx 0x7fc6f37fbf50 rcx 0x7fc7d9b66fb7 rdx 0x0 rdi 0x2 rsi 0x7fc6f37fbce0 rbp 0x7fc6f37fc050 rsp 0x7fc6f37fbce0 r8 0x0 r9 0x7fc6f37fbce0 r10 0x8 r11 0x246 r12 0x7fc6f37fbf50 r13 0x1000 r14 0x0 r15 0x30 rip 0x7fc7d9b66fb7 rflags 0x246 cs 0x33 fs 0x0 gs 0x0

We need to investigate on that, my feeling is you specified a very large id list and cause the issue. If you can range the ids, for example 5< pk <10 that make more sense to me.

**Comment by xiaofan-luan** [comment]
@yanliang567 could you help on verifying on that?
I doubt search with large PK id won't work

**Comment by ufuk23** [comment]
@xiaofan-luan It seems that the error is about Attu GUI because I try it with python code (pymilvus) on the same collection and don't have any error even 200K ids.

**Comment by yanliang567** [comment]
it seems that it only repo once, and it would succeed if try for more times.

**Comment by cydrain** [comment]
duplicated with #15574

**Comment by DragonDriver** [comment]
already fixed by #15663 
/assign @yanliang567
